hugo/content/guides/scoping/class-member.md

```diff
@@ -0,0 +1,105 @@
+---
+title: "Class Member Scoping"
+weight: 200
+---
+
+In this guide we will take at member based scoping. It's a mechanism you are likely familiar with from object oriented languages such as Java, C# and JavaScript:
+
+```java
+class A {
+    public B b;
+}
+
+class B {
+    public String value;
+}
+
+void test() {
+    A a = new A();
+    B b = a.b; // Refers to the `b` defined in class `A`
+    String value = b.value; // Refers to the `value` defined in class `B`
+}
+```
+
+Member based scoping like this requires not only a modification of the default scoping provider, but also some other prerequisites.
+This includes a member call mechanism in your grammar and a rudamentary type system.
+For this guide, we will use excerpts from the [langium-lox](https://github.com/langium/langium-lox) project.
+It implements a strongly-typed version of the [Lox language](https://craftinginterpreters.com/the-lox-language.html) from the popular book [Crafting Interpreters](https://craftinginterpreters.com/).
+
+We first start with the `MemberCall` grammar rule. It references one of our `NamedElements`, which could be variable declarations, functions, classes or methods and fields of those classes. Additionally, we allow function calls. Note that the grammar has no notion of whether these elements can actually be executed as functions. Instead, we always allow function calls on every named element and simply provide validation errors in case they were used erroneously. After parsing the first member call, we continue parsing further members as long as the input text provides us with further references to elements, separated by dots.
+
+```ts
+type NamedElement = FunctionDeclaration | VariableDeclaration | MethodMember | FieldMember | Class;
+
+MemberCall:
+    // Reference a named element of our grammar
+    // Variables, functions, etc.
+    element=[NamedElement:ID]
+    (explicitOperationCall?='(' (
+	    arguments+=Expression (',' arguments+=Expression)*
+	)? ')')?
+    // Create a new `MemberCall` and assign the old one to the `previous` property
+	({infer MemberCall.previous=current} 
+        // Member call with function call
+        ("." element=[NamedElement:ID] (
+		explicitOperationCall?='('
+		(
+		    arguments+=Expression (',' arguments+=Expression)*
+		)?
+		')')? 
+        // Chained function call without a member
+        | (
+		explicitOperationCall?='('
+		(
+		    arguments+=Expression (',' arguments+=Expression)*
+		)?
+		')'))
+    )*;
+```
+
+A very important aspect of these chained member calls is the action which rewrites the resulting AST. In this case, it effectively reverses the direction of member call AST nodes. Instead of starting with the first encountered member call and then traversing down to the last, we start with the last and traverse the list of member calls up using the `previous` property. The reason for doing this becomes clear when looking at the scope provider for the Lox language:
```

Comment:
```suggestion
A very important aspect of these chained member calls is the action which rewrites the resulting AST. In this case, it reverses the direction of member call AST nodes. Instead of starting with the first encountered member call and then traversing down to the last, we start with the last and traverse the list of member calls up using the `previous` property. The reason for doing this becomes clear when looking at the scope provider for the Lox language:
```